domaci_maze.py

```diff
@@ -90,495 +77,523 @@ def has_value(self) -> bool:
-        new_cell = self.board[new_row, new_col]
-        reward = new_cell.get_reward()
-        is_terminal = new_cell.is_terminal()
-        return (new_row, new_col), reward, is_terminal
-
-    def get_states(self):
-        states = []
-        for r in range(self.board.rows_no):
-            for c in range(self.board.cols_no):
-                if self.board[r, c].is_steppable():
-                    states.append((r, c))
-        return states
-
-    def is_terminal(self, s):
-        return self.board[s[0], s[1]].is_terminal()
-
-    def get_actions(self, cell):
-        if cell in self.graph:
-            return list(self.graph[cell].keys())
-        else:
+    def set_probabilities(self, node, g):
+
+        # terminal and wall node do not have any probabilities
+        if isinstance(node, WallNode) or isinstance(node, TerminalNode):
             return []
 
+        # list is easier to iterate, only because of that is node_list made
+        nodes_list = [node for node in g]
+        nodes_list_help = copy(nodes_list)  # only for iteration
+
+        probabilities = []
+
+        total_cells = self.graph_width * self.graph_height
+        zero_cells = total_cells // 2  # number of cells with zero prob of stepping
+
+        # probability is set to zero for wall modes
+        for n in nodes_list_help:
+            if isinstance(n, WallNode):
+                probabilities.append([0, n, 0])
+                zero_cells -= 1
+                nodes_list.remove(n)
+
+        # for regular nodes, half of probabilities should be zero, including wall nodes
+        if isinstance(node, RegularNode):
+            while zero_cells != 0:
+                random_node = rdm.choice(nodes_list)
+                nodes_list.remove(random_node)
+                probabilities.append([0, random_node, 0])
+                zero_cells -= 1
+
+        # number of nodes that can be reached with one action
+        # action_len = math.ceil(len(nodes_list) / len(ACTIONS))
+        if len(nodes_list) / len(ACTIONS) > 1.5:
+            action_len = math.ceil(len(nodes_list) / len(ACTIONS))
+        else:
+            action_len = math.floor(len(nodes_list) / len(ACTIONS))
+
+        # shuffle actions
+        random_actions = ACTIONS
+        rdm.shuffle(random_actions)
+        actions_copy = copy(random_actions)  # actions yet to be processed
 
-def create_graph(maze_env):
-    G = nx.DiGraph()
-    for cell, actions in maze_env.graph.items():
-        for action, target in actions.items():
-            source = (
-                cell.row,
-                cell.col,
-            )  # assuming each cell has row and col attributes
-            G.add_edge(source, target, action=action)
-    return G
+        for action in random_actions:
+
+            #  when it's the last action, all the left nodes are connected to the last action
+            if len(actions_copy) == 1:
+                nodes_for_current_action = nodes_list

```

Comment:
Umesto ovoga si samo mogao reci for action in random_actions[:1] i onda posle fora uzeti i dodeliti poslednjoj akciji sve sta treba. Ovako bezveze zauzmes dosta memorije za celu novu listu

Reply:
Generalno da, ali to za memoriju nema veze jer je ovo lista od par elemenata, ne znaci nista

Reply:
Da se ne zbunimo na odbrani zasto to postoji:)